zingolib/src/wallet/notes/sapling.rs

```diff
@@ -171,3 +171,93 @@ impl ShieldedNoteInterface for SaplingNote {
+        wallet::{notes::ShieldedNoteInterface, traits::FromBytes},
+    };
+
+    use super::SaplingNote;
+
+    /// builds a mock transparent note after all pieces are supplied
+    pub(crate) struct SaplingNoteBuilder {
+        diversifier: Option<sapling_crypto::Diversifier>,
+        note: Option<sapling_crypto::Note>,
+        witnessed_position: Option<Option<Position>>,
+        output_index: Option<Option<u32>>,
+        nullifier: Option<Option<sapling_crypto::Nullifier>>,
+        spent: Option<Option<(TxId, u32)>>,
+        unconfirmed_spent: Option<Option<(TxId, u32)>>,
+        memo: Option<Option<Memo>>,
+        is_change: Option<bool>,
+        have_spending_key: Option<bool>,
+    }
+
+    #[allow(dead_code)] //TODO:  fix this gross hack that I tossed in to silence the language-analyzer false positive
+    impl SaplingNoteBuilder {
+        pub fn new() -> Self {
+            Self::default()
+        }
+
+        // Methods to set each field
+        build_method!(diversifier, sapling_crypto::Diversifier);
+        build_method!(note, sapling_crypto::Note);
+        build_method!(witnessed_position, Option<Position>);
+        build_method!(output_index, Option<u32>);
+        build_method!(nullifier, Option<sapling_crypto::Nullifier>);
+        build_method!(spent, Option<(TxId, u32)>);
+        build_method!(unconfirmed_spent, Option<(TxId, u32)>);
+        build_method!(memo, Option<Memo>);
+        pub fn set_change(mut self, is_change: bool) -> Self {
+            self.is_change = Some(is_change);
+            self
+        }
+        build_method!(have_spending_key, bool);
+
+        // Build method
+        pub fn build(self) -> SaplingNote {
+            SaplingNote::from_parts(
+                self.diversifier.unwrap(),
+                self.note.unwrap(),
+                self.witnessed_position.unwrap(),
+                self.nullifier.unwrap(),
+                self.spent.unwrap(),
+                self.unconfirmed_spent.unwrap(),
+                self.memo.unwrap(),
+                self.is_change.unwrap(),
+                self.have_spending_key.unwrap(),
+                self.output_index.unwrap(),
+            )
+        }
+    }
+
+    impl Default for SaplingNoteBuilder {
+        fn default() -> Self {
+            SaplingNoteBuilder {
+                diversifier: Some(sapling_crypto::Diversifier([0; 11])),
+                note: Some(crate::test_framework::mocks::mock_sapling_crypto_note()),
+                witnessed_position: Some(Some(Position::from(0))),
+                output_index: Some(Some(0)),
+                nullifier: Some(Some(sapling_crypto::Nullifier::from_bytes([0; 32]))),
+                spent: Some(None),
+                unconfirmed_spent: Some(None),
+                memo: Some(None),
+                is_change: Some(false),
+                have_spending_key: Some(true),
+            }
+        }
+    }
+
+    impl SaplingNote {
+        #[allow(dead_code)]
```

Comment:
i wonder if it would be better to remove all these allow dead code cfgs and instead make all these mocks `pub` so we can use them across all our crates?